core/src/test/java/io/questdb/test/griffin/engine/functions/groupby/FirstNotNullGroupByFunctionFactoryTest.java

```diff
@@ -0,0 +1,120 @@
+                        "from long_sequence(3))",
+                null,
+                false,
+                true
+        );
+    }
+
+    @Test
+    public void testFirstNotNull() throws Exception {
+
+        ddl("create table tab (a0 char,a1 date,a2 double,a3 float,a4 int,a5 long,a6 symbol,a7 timestamp,a8 uuid,a9 string)");
+
+        long now = Instant.now().truncatedTo(ChronoUnit.DAYS).getEpochSecond() * 1_000_000;
+        UUID firstUuid = UUID.randomUUID();
+        try (TableWriter w = getWriter("tab")) {
+            for (int i = 10; i > 0; i--) {
+                TableWriter.Row r = w.newRow();
+                r.append();
+            }
+            w.commit();
+            TableWriter.Row r = w.newRow();
+            r.putChar(0, 'a');
+            r.putDate(1, now);
+            r.putDouble(2, 2.2);
+            r.putFloat(3, 3.3f);
+            r.putInt(4, 4);
+            r.putLong(5, 5L);
+            r.putSym(6, "a_symbol");
+            r.putTimestamp(7, now + 700);
+            r.putUuid(8, firstUuid.toString());
+            r.putStr(9, "a_string");
+            r.append();
+
+            r = w.newRow();
+            r.putChar(0, 'b');
+            r.putDate(1, now + 1000);
+            r.putDouble(2, 22.2);
+            r.putFloat(3, 33.3f);
+            r.putInt(4, 44);
+            r.putLong(5, 55L);
+            r.putSym(6, "b_symbol");
+            r.putTimestamp(7, now + 1700);
+            r.putUuid(8, UUID.randomUUID().toString());
+            r.putStr(9, "b_string");
+            r.append();
+
+            w.commit();
+        }
+
+        try (RecordCursorFactory factory = select(
+                "select first_not_null(a0)a0,first_not_null(a1)a1,first_not_null(a2)a2,first_not_null(a3)a3,first_not_null(a4)a4," +
+                        "first_not_null(a5)a5,first_not_null(a6)a6,first_not_null(a7)a7,first_not_null(a8)a8,first_not_null(a9)a9 from tab")) {
+            try (RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
+                Record record = cursor.getRecord();
+                Assert.assertEquals(1, cursor.size());
+                Assert.assertTrue(cursor.hasNext());
+
+                Assert.assertEquals('a', record.getChar(0));
+                Assert.assertEquals(now, record.getDate(1));
+                Assert.assertEquals(2.2, record.getDouble(2), 0.001);
+                Assert.assertEquals(3.3, record.getFloat(3), 0.001);
+                Assert.assertEquals(4, record.getInt(4));
+                Assert.assertEquals(5, record.getLong(5));
+                Assert.assertEquals("a_symbol", record.getSym(6));
+                Assert.assertEquals(now + 700, record.getTimestamp(7));
+                Assert.assertEquals(firstUuid, new UUID(record.getLong128Hi(8), record.getLong128Lo(8)));
+                Assert.assertEquals("a_string", record.getStr(9).toString());
```

Comment:
nit: we have `TestUtils#assertEquals()` for comparisons around `CharSequence`s.

Reply:
updated